File: weld_engine/calculators.py
"""
WeldScan 3D — Welder-Specific Math Calculators

Pure-Python implementations of the welding math formulas.
These functions have no OpenCV dependency and can be unit-tested standalone.
"""

import math


def bend_deduction(inside_radius_mm: float, thickness_mm: float) -> float:
    """
    Bend deduction (BD) — flat material consumed by a bend.

    Formula: BD = (π/2) × (inside_radius + thickness / 3)

    Args:
        inside_radius_mm: inside bend radius in mm
        thickness_mm:     metal thickness in mm

    Returns:
        Bend deduction in mm
    """
    if inside_radius_mm < 0:
        raise ValueError("inside_radius_mm must be non-negative")
    if thickness_mm <= 0:
        raise ValueError("thickness_mm must be positive")
    return (math.pi / 2) * (inside_radius_mm + thickness_mm / 3)


def bevel_gap(wire_diameter_mm: float, thickness_mm: float) -> dict:
    """
    Bevel and root-gap calculator (AWS D1.1 rule of thumb).

    Args:
        wire_diameter_mm: electrode/wire diameter in mm
        thickness_mm:     metal thickness in mm

    Returns:
        dict: {root_gap_mm, landing_mm}
    """
    if wire_diameter_mm <= 0:
        raise ValueError("wire_diameter_mm must be positive")
    if thickness_mm <= 0:
        raise ValueError("thickness_mm must be positive")
    root_gap_mm = wire_diameter_mm
    landing_mm = max(1.0, thickness_mm * 0.1)
    return {"root_gap_mm": round(root_gap_mm, 2), "landing_mm": round(landing_mm, 2)}


def fillet_cross_section(leg_mm: float) -> float:
    """
    Cross-section area of an equal-leg fillet weld.

    A = 0.5 × leg²

    Args:
        leg_mm: fillet leg size in mm

    Returns:
        Cross-section area in mm²
    """
    if leg_mm <= 0:
        raise ValueError("leg_mm must be positive")
    return 0.5 * leg_mm * leg_mm


def weld_volume(cross_section_mm2: float, weld_length_mm: float) -> dict:
    """
    Estimate weld volume for wire/gas consumption calculations.

    V = cross_section_area × weld_length

    Args:
        cross_section_mm2: weld cross-section area in mm²
        weld_length_mm:    total weld run length in mm

    Returns:
        dict: {volume_mm3, volume_cm3, volume_in3}
    """
    if cross_section_mm2 <= 0:
        raise ValueError("cross_section_mm2 must be positive")
    if weld_length_mm <= 0:
        raise ValueError("weld_length_mm must be positive")
    volume_mm3 = cross_section_mm2 * weld_length_mm
    volume_cm3 = volume_mm3 / 1000.0
    volume_in3 = volume_mm3 / 16387.064
    return {
        "volume_mm3": round(volume_mm3, 4),
        "volume_cm3": round(volume_cm3, 4),
        "volume_in3": round(volume_in3, 4),
    }

File: src/weldscan/calculators.js
/**
 * WeldScan 3D — Welder-Specific Math Calculators
 *
 * Pure JavaScript implementation of the welding math engine.
 * These functions are used by the Express API routes and can also
 * be unit-tested independently of any CV pipeline.
 */

/**
 * Compute the pixels-per-inch (PPI) calibration constant.
 *
 * @param {number} refPixelLength  - Measured length of the reference object in pixels
 * @param {number} refRealInches   - Known real-world length of the reference object (inches)
 * @returns {number} PPI ratio
 */
function computePPI(refPixelLength, refRealInches) {
  if (refPixelLength <= 0) throw new Error('refPixelLength must be positive');
  if (refRealInches <= 0) throw new Error('refRealInches must be positive');
  return refPixelLength / refRealInches;
}

/**
 * Convert a pixel distance to inches using the calibration PPI.
 *
 * @param {number} pixels
 * @param {number} ppi
 * @returns {number} distance in inches
 */
function pixelsToInches(pixels, ppi) {
  if (ppi <= 0) throw new Error('ppi must be positive');
  return pixels / ppi;
}

/**
 * Convert inches to millimetres.
 *
 * @param {number} inches
 * @returns {number}
 */
function inchesToMm(inches) {
  return inches * 25.4;
}

/**
 * Compute the angle (degrees) between two 2-D vectors formed by
 * three consecutive contour points: A → B → C.
 *
 * @param {{x:number, y:number}} A
 * @param {{x:number, y:number}} B  - vertex
 * @param {{x:number, y:number}} C
 * @returns {number} interior angle at B in degrees
 */
function angleBetween(A, B, C) {
  const BAx = A.x - B.x;
  const BAy = A.y - B.y;
  const BCx = C.x - B.x;
  const BCy = C.y - B.y;
  const dot = BAx * BCx + BAy * BCy;
  const magBA = Math.hypot(BAx, BAy);
  const magBC = Math.hypot(BCx, BCy);
  if (magBA === 0 || magBC === 0) return 0;
  const cosTheta = Math.max(-1, Math.min(1, dot / (magBA * magBC)));
  return (Math.acos(cosTheta) * 180) / Math.PI;
}

/**
 * Compute the miter-cut angle needed for a chop saw.
 * Miter angle = (180 - interior_angle) / 2
 *
 * @param {number} interiorDeg - interior angle between two segments (degrees)
 * @returns {number} miter cut angle in degrees
 */
function miterAngle(interiorDeg) {
  return (180 - interiorDeg) / 2;
}

/**
 * Kerf-adjusted segment length.
 * Each cut removes `kerfMm` of material; shift both ends inward by half.
 *
 * @param {number} nominalLengthMm  - nominal segment length in mm
 * @param {number} kerfMm           - kerf width in mm (tool-specific)
 * @returns {number} adjusted length in mm
 */
function kerfAdjustedLength(nominalLengthMm, kerfMm) {
  if (kerfMm < 0) throw new Error('kerfMm must be non-negative');
  return nominalLengthMm - kerfMm;
}

/**
 * Bend deduction (BD) — the amount of flat material "consumed" by a bend.
 * Formula: BD = (π/2) × (insideRadius + thickness / 3)
 *
 * @param {number} insideRadiusMm  - inside bend radius in mm
 * @param {number} thicknessMm     - material thickness in mm
 * @returns {number} bend deduction in mm
 */
function bendDeduction(insideRadiusMm, thicknessMm) {
  if (insideRadiusMm < 0) throw new Error('insideRadiusMm must be non-negative');
  if (thicknessMm <= 0) throw new Error('thicknessMm must be positive');
  return (Math.PI / 2) * (insideRadiusMm + thicknessMm / 3);
}

/**
 * Bevel and root-gap calculator.
 * Given wire diameter and material thickness, returns:
 *   - rootGapMm: required root opening
 *   - landingMm: required root face (landing)
 *
 * Rule-of-thumb (AWS D1.1):
 *   rootGap  ≈ wireDiameter
 *   landing  ≈ max(1, thicknessMm * 0.1)  (at least 1 mm, up to 10% of thickness)
 *
 * @param {number} wireDiameterMm
 * @param {number} thicknessMm
 * @returns {{ rootGapMm: number, landingMm: number }}
 */
function bevelGap(wireDiameterMm, thicknessMm) {
  if (wireDiameterMm <= 0) throw new Error('wireDiameterMm must be positive');
  if (thicknessMm <= 0) throw new Error('thicknessMm must be positive');
  const rootGapMm = wireDiameterMm;
  const landingMm = Math.max(1, thicknessMm * 0.1);
  return { rootGapMm, landingMm };
}

/**
 * Weld volume estimator.
 * V = cross_section_area × weld_length
 *
 * Cross-section area for a fillet weld: A = 0.5 × leg × leg
 * For a butt/groove weld: A = thickness × (rootGap + 0.5 × bevelAngleRad × thickness)
 *
 * This function accepts a pre-computed cross-section area for flexibility.
 *
 * @param {number} crossSectionMm2  - weld cross-section area in mm²
 * @param {number} weldLengthMm     - total weld run length in mm
 * @returns {{ volumeMm3: number, volumeCm3: number, volumeIn3: number }}
 */
function weldVolume(crossSectionMm2, weldLengthMm) {
  if (crossSectionMm2 <= 0) throw new Error('crossSectionMm2 must be positive');
  if (weldLengthMm <= 0) throw new Error('weldLengthMm must be positive');
  const volumeMm3 = crossSectionMm2 * weldLengthMm;
  const volumeCm3 = volumeMm3 / 1000;
  const volumeIn3 = volumeMm3 / 16387.064;
  return { volumeMm3, volumeCm3, volumeIn3 };
}

/**
 * Fillet weld cross-section area (equal-leg fillet).
 *
 * @param {number} legMm - fillet leg size in mm
 * @returns {number} area in mm²
 */
function filletCrossSection(legMm) {
  if (legMm <= 0) throw new Error('legMm must be positive');
  return 0.5 * legMm * legMm;
}

/**
 * Build a cut list from an array of polygon segments.
 *
 * @param {Array<{lengthPx: number, angle: number}>} segments
 *   - lengthPx: segment length in pixels
 *   - angle: miter angle in degrees for this segment's start cut
 * @param {number} ppi       - pixels per inch
 * @param {number} kerfMm    - kerf width in mm (0 = no adjustment)
 * @returns {Array<{label:string, lengthIn:number, lengthMm:number, miterDeg:number, kerfAdjLengthMm:number}>}
 */
function buildCutList(segments, ppi, kerfMm = 0) {
  return segments.map((seg, i) => {
    const lengthIn = pixelsToInches(seg.lengthPx, ppi);
    const lengthMm = inchesToMm(lengthIn);
    const kerfAdjLengthMm = kerfAdjustedLength(lengthMm, kerfMm);
    return {
      label: `P${i + 1}`,
      lengthIn: +lengthIn.toFixed(3),
      lengthMm: +lengthMm.toFixed(2),
      miterDeg: +seg.angle.toFixed(1),
      kerfAdjLengthMm: +kerfAdjLengthMm.toFixed(2),
    };
  });
}

module.exports = {
  computePPI,
  pixelsToInches,
  inchesToMm,
  angleBetween,
  miterAngle,
  kerfAdjustedLength,
  bendDeduction,
  bevelGap,
  weldVolume,
  filletCrossSection,
  buildCutList,
};
